### ISSUE TITLE
Rename `usingnamespace` to `mixin`

### ISSUE THREAD

**Issue by ifreund** [ziglang/zig#9861]
The semantics of `usingnamespace` were changed in #9618 as per proposal #9629. However, the keyword itself was not changed and it was [decided](https://github.com/ziglang/zig/issues/9629#issuecomment-911411325) to leave that for a separate proposal.

To reiterate my [comment](https://github.com/ziglang/zig/issues/9629#issuecomment-907146993) on #9629, I think the originally proposed `includenamespace` isn't perfect as not everything in the target namespace is included, only the public declarations.

I think using `mixin` as the new keyword would be more clear and concise. Mixins are already a well known concept in programming ([wikipedia](https://en.wikipedia.org/wiki/Mixin)) and the revised `usingnamespace` semantics map quite well to the generally understood concept of a mixin.

I also proposed using `includedecls` or `includedeclarations` in the original thread, however I find these ugly compared to `mixin` and don't think they give a clearer idea of the semantics.

As this is a rather zig-specific concept, people new to zig will almost certainly need to read the documentation to learn the exact semantics of this keyword no matter what we choose.

**Comment by nmichaels** [comment]
I like `mixin` better than any of the alternatives you listed, but it implies something false to me. When I see `mixin` it makes me think that the mixed-in things will be available in the local namespace. `reexport` is awkward but closer to the intent. `transship`?

This feels a little bit bike-sheddy, but the feature's implemented and good names are really important. If we take a name for a concept with which people are familiar (like `using namespace`) then users will be surprised when it has different semantics from what they expect. If it's an entirely unfamiliar name, they'll have to look up what it does, which is probably what we want. The balance, I think, is finding a name that's unfamiliar enough to cause transplants from other languages to look it up, while being intuitive enough that Zig programmers can remember it easily.

**Comment by InKryption** [comment]
More bike-shedding here, but how about `forward`, `forwarddecls` or `forwardnamespace`? Being that the new semantics are closer to forwarding declarations, as opposed to mixing them in with the present declarations, or including them.

**Comment by rohlem** [comment]
I've never used a language with a `mixin` keyword, but Wikipedia says many languages do different things with it, so I'm not convinced that the word implies influencing identifier lookup. They _are_ available in the namespace, via indirection of `@This()`.

To explicitly show only `pub` things are included, we could put it in the name: `includepub` / `mixinpub` .

**Comment by maxxnino** [comment]
addon, plugin, extend 🙄

**Comment by zzyxyzz** [comment]
+1 for `forwarddecls`. It's kinda ugly, but describes the semantics precisely. Since this is not a frequently used feature, a long and descriptive keyword should be better than a short but somewhat cryptic one.

**Comment by kenaryn** [comment]
What about ```nonlocal``` or ```outerspace```?

**Comment by jibal** [comment]
How about calling it mixindeclarations ? Or mixindeclarationsof ?

**Comment by Hsad** [comment]
I like the logic of `adopt`, it implies the superseding behavior of needing to label the namespace to access it.  Sort of like a new family name.  Something like `AdoptNames`, or `AdoptDef` could do a better job of hinting towards the actual use case.   Usurp, Clone, Sample, Inherit, Use, Reference/Refencing, Take, Utilize, + Names/Def, etc, could be other variations with a similar train of thought.

**Comment by nektro** [comment]
I like `usingnamespace` being long, and "namespace" being in the name accurately reflects that only the decls are being imported

**Comment by jibal** [comment]
But the decls aren't imported ... not any more; they are added to the container and available to its users, but they aren't available to the code within it.

**Comment by igor84** [comment]
Based on the comments here I couldn't quite grasp why `mixin` is not a perfect choice, but after playing around with it this example made me realize:

```zig
fn SA(comptime Self: type) type {
    return struct {
        a: u8 = 80,

        pub fn add(self: Self, b: u8) u8 {
            return self.a + b;
        }
    };
}

const SB = struct {
    c: u8 = 80,
    pub usingnamespace SA(@This());
};
```

This example produces a compile error `no member 'a' in struct 'SB'` on `return self.a + b;`. That is because `usingnamespace` only includes the declarations (functions and consts) but not fields from `SA` into `SB`. Note that if you changed that line to `return self.c + b;` it would compile since `Self` in `SA` will actually refer to `SB`.

To me the word `mixin` implies mixing in fields should also work since I first learned about the concept in DLang. But, wikipedia is a bit confusion on the terms. The [mixin](https://en.wikipedia.org/wiki/Mixin) page actually describes `mixin` as something that only mixes in additional methods which actually is what is happening here. On the other hand the page about [trait](https://en.wikipedia.org/wiki/Trait_(computer_programming)) specifically describes the difference from mixing with this sentence:

> In contrast, mixins include full method definitions and may also carry state through member variable, while traits usually don't.

I still like the choice of `mixin` keyword the most but I do understand now why it is not perfect.

**Comment by daneelsan** [comment]
Because I didn't see this above I'll propose: `usingdeclarations`.

**Comment by ominitay** [comment]
`forwarddecls`, or `forwarddeclarations`, as suggested by @InKryption more adequately describes the behaviour than `usingdeclarations`

**Comment by nektro** [comment]
`namespace` is already the name used internally for the list of a container's declarations. the name is fine

**Comment by ominitay** [comment]
@nektro The term `using` isn't great, can be misleading in this context though. This is obviously nitpicking though lol.

**Comment by jibal** [comment]
My suggestion of `mixindeclarations` got a lot of downvotes even though it is more accurate than many of the other proposals and clearer and less ambiguous than simply "mixin" which is what this issue proposes.

**Comment by codethief** [comment]
> but on reconsideration I think that `forwarddeclarations` is better ... `mixin`, like `using`, has the wrong implication as @nmichaels explained above.

I agree with @jibal's assessment here. In fact, it took me half an hour now (of reading the docs as well as the various Github issues linked here) to realize that, contrary to its name, `usingnamespace` does _not_ (or no longer, as of #9629) allow using the members of the given namespace in the current scope (the one that contains the `usingnamespace` instruction). This is _very_ confusing.

**Comment by jibal** [comment]
> rather it's merging the declarations from it

But it's not ... please see the discussion above. The keyword has very odd semantics ... it exports the symbols from the referenced struct but it doesn't import them ... the code containing the keyword cannot use the symbols. That's the whole reason for this proposed change.

In any case I can't see how "derive" or "deriving" give any indication of what it does, and suggests a relationship that may not hold at all.

**Comment by zzyxyzz** [comment]
Since we're still bikeshedding, how about `mixout`?
It's a `mixin` that only has an effect outside the current namespace ))
Also unusual enough to hit the docs right away, without forming wrong expectations first.

**Comment by deflock** [comment]
`pollute` 🤪

**Comment by jibal** [comment]
None of these suggestions is clearly better than `forwarddecls` (or `forwarddeclarations` -- there's no need to be cryptic with this rarely used keyword), suggested over 2 years ago.

**Comment by mlugg** [comment]
`derive` would be pretty misleading. It gives implications of OO-like behaviours: inheriting fields, and adjusting method types to operate on this type. In reality, all the keyword does is directly include all symbols from another namespace.

**Comment by jibal** [comment]
"derive" is wrong in every possible way.

**Comment by jedisct1** [comment]
How about "merge"?

**Comment by zzyxyzz** [comment]
> How about "merge"?

That would not be a good keyword to reserve, IMHO.

**Comment by nektro** [comment]
usingnamespace is good as is

**Comment by deflock** [comment]
Okay, brainstorming is going on: `involve`, `mixup`, or some re-prefixed `reintroduce`, `reuse`, `reinject`

**Comment by hsyl20** [comment]
`reexport`

**Comment by joelreymont** [comment]
`mixin`.
